File: experiment-5/runs/language=go_model=claude-opus-4-8_tooling=none/rep1/internal/soccer/format.go
// Context:
//   - This file renders query results as the human-readable text the MCP tools
//     return to the LLM/user. The formats mirror the example answer layouts in
//     the specification (match lists with date + score + competition, team
//     records with win rate, standings tables, player rankings, etc.).
//   - Keeping formatting here (rather than in the MCP layer) lets the same
//     rendering be unit-tested directly and keeps the tool handlers thin.
package soccer

import (
	"fmt"
	"strings"
)

// FormatDate renders a match date, or "date unknown" when absent.
func (m Match) FormatDate() string {
	if !m.HasDate {
		return "date unknown"
	}
	return m.Date.Format("2006-01-02")
}

// Line renders a single match as one summary line.
func (m Match) Line() string {
	score := "vs"
	if m.HasScore {
		score = fmt.Sprintf("%d-%d", m.HomeGoals, m.AwayGoals)
	}
	ctx := m.Competition
	switch {
	case m.Round != "":
		ctx += " Round " + m.Round
	case m.Stage != "":
		ctx += " - " + m.Stage
	}
	return fmt.Sprintf("%s: %s %s %s (%s)", m.FormatDate(), m.HomeTeam, score, m.AwayTeam, ctx)
}

// FormatMatches renders a list of matches with an optional title and a cap on
// how many lines to print (the rest are summarised as "... (N more)").
func FormatMatches(title string, matches []Match, max int) string {
	var b strings.Builder
	if title != "" {
		b.WriteString(title)
		b.WriteString("\n")
	}
	if len(matches) == 0 {
		b.WriteString("No matches found.")
		return b.String()
	}
	shown := matches
	if max > 0 && len(matches) > max {
		shown = matches[:max]
	}
	for _, m := range shown {
		b.WriteString("- ")
		b.WriteString(m.Line())
		b.WriteString("\n")
	}
	if len(matches) > len(shown) {
		fmt.Fprintf(&b, "... (%d more match(es))\n", len(matches)-len(shown))
	}
	fmt.Fprintf(&b, "\nTotal: %d match(es).", len(matches))
	return b.String()
}

// FormatHeadToHead renders a head-to-head summary.
func (h HeadToHead) Format(maxMatches int) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Head-to-head: %s vs %s\n", h.Team1, h.Team2)
	if len(h.Matches) == 0 {
		b.WriteString("No matches found between these teams.")
		return b.String()
	}
	fmt.Fprintf(&b, "Record: %s %d wins, %s %d wins, %d draws\n",
		h.Team1, h.Team1Wins, h.Team2, h.Team2Wins, h.Draws)
	fmt.Fprintf(&b, "Goals: %s %d, %s %d\n\n", h.Team1, h.Team1Goals, h.Team2, h.Team2Goals)
	b.WriteString(FormatMatches("Matches:", h.Matches, maxMatches))
	return b.String()
}

// Format renders a team record.
func (r TeamRecord) Format() string {
	var b strings.Builder
	header := r.Team + " record"
	var quals []string
	if r.Season != 0 {
		quals = append(quals, fmt.Sprintf("%d", r.Season))
	}
	if r.Competition != "" {
		quals = append(quals, canonicalCompetition(r.Competition))
	}
	if r.Venue != "" {
		quals = append(quals, strings.ToLower(r.Venue))
	}
	if len(quals) > 0 {
		header += " (" + strings.Join(quals, " ") + ")"
	}
	b.WriteString(header + ":\n")
	fmt.Fprintf(&b, "- Matches: %d\n", r.Played)
	fmt.Fprintf(&b, "- Wins: %d, Draws: %d, Losses: %d\n", r.Wins, r.Draws, r.Losses)
	fmt.Fprintf(&b, "- Goals For: %d, Goals Against: %d (diff %+d)\n", r.GoalsFor, r.GoalsAgainst, r.GoalDiff())
	fmt.Fprintf(&b, "- Points: %d\n", r.Points())
	fmt.Fprintf(&b, "- Win rate: %.1f%%", r.WinRate()*100)
	return b.String()
}

// FormatStandings renders a computed league table.
func FormatStandings(competition string, season int, table []Standing, max int) string {
	var b strings.Builder
	fmt.Fprintf(&b, "%s %d Standings (calculated from matches):\n", canonicalCompetition(competition), season)
	if len(table) == 0 {
		b.WriteString("No results available for this competition/season.")
		return b.String()
	}
	shown := table
	if max > 0 && len(table) > max {
		shown = table[:max]
	}
	for _, row := range shown {
		fmt.Fprintf(&b, "%2d. %s - %d pts (%dW %dD %dL, GD %+d)\n",
			row.Position, row.Team, row.Points(), row.Wins, row.Draws, row.Losses, row.GoalDiff())
	}
	if len(table) > len(shown) {
		fmt.Fprintf(&b, "... (%d more)\n", len(table)-len(shown))
	}
	if len(table) > 0 {
		fmt.Fprintf(&b, "\nChampion: %s", table[0].Team)
	}
	return b.String()
}

// Format renders league aggregate statistics.
func (l LeagueStats) Format() string {
	var b strings.Builder
	scope := canonicalCompetition(l.Competition)
	if l.Season != 0 {
		scope = fmt.Sprintf("%s %d", scope, l.Season)
	}
	fmt.Fprintf(&b, "%s statistics:\n", scope)
	fmt.Fprintf(&b, "- Matches: %d\n", l.Matches)
	fmt.Fprintf(&b, "- Total goals: %d\n", l.TotalGoals)
	fmt.Fprintf(&b, "- Average goals per match: %.2f\n", l.AvgGoals())
	fmt.Fprintf(&b, "- Home wins: %d (%.1f%%), Away wins: %d, Draws: %d\n",
		l.HomeWins, l.HomeWinRate()*100, l.AwayWins, l.Draws)
	if len(l.BiggestWins) > 0 {
		b.WriteString("\nBiggest wins:\n")
		for i, m := range l.BiggestWins {
			fmt.Fprintf(&b, "%d. %s\n", i+1, m.Line())
		}
	}
	return strings.TrimRight(b.String(), "\n")
}

// Line renders a single player as one summary line.
func (p Player) Line() string {
	parts := []string{p.Name}
	if p.Overall > 0 {
		parts = append(parts, fmt.Sprintf("Overall: %d", p.Overall))
	}
	if p.Position != "" {
		parts = append(parts, "Position: "+p.Position)
	}
	if p.Club != "" {
		parts = append(parts, "Club: "+p.Club)
	}
	if p.Nationality != "" {
		parts = append(parts, p.Nationality)
	}
	return strings.Join(parts, " - ")
}

// FormatPlayers renders a ranked player list.
func FormatPlayers(title string, players []Player, max int) string {
	var b strings.Builder
	if title != "" {
		b.WriteString(title + "\n")
	}
	if len(players) == 0 {
		b.WriteString("No players found.")
		return b.String()
	}
	shown := players
	if max > 0 && len(players) > max {
		shown = players[:max]
	}
	for i, p := range shown {
		fmt.Fprintf(&b, "%d. %s\n", i+1, p.Line())
	}
	if len(players) > len(shown) {
		fmt.Fprintf(&b, "... (%d more player(s))\n", len(players)-len(shown))
	}
	fmt.Fprintf(&b, "\nTotal: %d player(s).", len(players))
	return b.String()
}

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=beads/rep3/src/format.ts
/*
 * ============================================================================
 * Context
 * ----------------------------------------------------------------------------
 * Module:  src/format.ts
 * Purpose: Render query results into the human-readable text shown in the spec
 *          ("Flamengo 2-1 Fluminense (Brasileirão Round 22)", standings tables,
 *          player lists, etc.). MCP tools return both this text and structured
 *          JSON; the text is what an LLM/end-user reads.
 * Inputs:  Query-layer result objects.
 * Outputs: Formatted strings.
 * Notes:   Pure formatting only — no data access.
 * ============================================================================
 */

import type { Match, Player, StandingRow } from "./data/types.js";
import type { HeadToHead } from "./queries/matches.js";
import type { TeamRecord } from "./data/types.js";

/** Format a single match as a one-line summary. */
export function formatMatch(m: Match): string {
  const date = m.date ?? "????-??-??";
  const score =
    m.homeGoals != null && m.awayGoals != null
      ? `${m.homeGoals}-${m.awayGoals}`
      : "vs";
  const ctx: string[] = [m.competition];
  if (m.round) ctx.push(`Round ${m.round}`);
  if (m.stage) ctx.push(m.stage);
  return `${date}: ${m.homeTeam} ${score} ${m.awayTeam} (${ctx.join(", ")})`;
}

/** Format a list of matches with an optional cap and "N more" note. */
export function formatMatchList(matches: Match[], cap = 25): string {
  if (matches.length === 0) return "No matches found.";
  const shown = matches.slice(0, cap);
  const lines = shown.map((m) => `- ${formatMatch(m)}`);
  if (matches.length > cap) {
    lines.push(`- ... (${matches.length - cap} more matches in dataset)`);
  }
  return lines.join("\n");
}

/** Format a head-to-head summary block. */
export function formatHeadToHead(h: HeadToHead): string {
  const header = `${h.team1} vs ${h.team2} — head-to-head (dataset):`;
  const record =
    `Total matches: ${h.totalMatches} | ` +
    `${h.team1} ${h.team1Wins} wins, ${h.team2} ${h.team2Wins} wins, ${h.draws} draws | ` +
    `Goals: ${h.team1} ${h.team1Goals} - ${h.team2Goals} ${h.team2}`;
  const list = formatMatchList(h.matches, 10);
  return `${header}\n${record}\n\nRecent meetings:\n${list}`;
}

/** Format a team's W/D/L + goals record. */
export function formatTeamRecord(rec: TeamRecord, scope = ""): string {
  const rate =
    rec.played === 0
      ? "0.0%"
      : `${((rec.wins / rec.played) * 100).toFixed(1)}%`;
  const title = scope ? `${rec.team} record (${scope}):` : `${rec.team} record:`;
  return [
    title,
    `- Matches: ${rec.played}`,
    `- Wins: ${rec.wins}, Draws: ${rec.draws}, Losses: ${rec.losses}`,
    `- Goals For: ${rec.goalsFor}, Goals Against: ${rec.goalsAgainst}`,
    `- Points: ${rec.points}`,
    `- Win rate: ${rate}`,
  ].join("\n");
}

/** Format a league standings table. */
export function formatStandings(
  rows: StandingRow[],
  title: string,
  cap = 30,
): string {
  if (rows.length === 0) return `No data to compute standings for ${title}.`;
  const lines = [`${title} (calculated from matches):`];
  const shown = rows.slice(0, cap);
  for (const r of shown) {
    const tag =
      r.position === 1 ? " - Champion" : "";
    lines.push(
      `${r.position}. ${r.team} - ${r.points} pts ` +
        `(${r.wins}W, ${r.draws}D, ${r.losses}L, GF ${r.goalsFor}, GA ${r.goalsAgainst}, GD ${r.goalDifference >= 0 ? "+" : ""}${r.goalDifference})${tag}`,
    );
  }
  return lines.join("\n");
}

/** Format a single player line. */
export function formatPlayer(p: Player, index?: number): string {
  const prefix = index != null ? `${index}. ` : "";
  const bits = [
    `Overall: ${p.overall ?? "?"}`,
    `Position: ${p.position || "?"}`,
    `Club: ${p.club || "Free agent"}`,
    `Nationality: ${p.nationality || "?"}`,
  ];
  if (p.age != null) bits.push(`Age: ${p.age}`);
  return `${prefix}${p.name} - ${bits.join(", ")}`;
}

/** Format a list of players. */
export function formatPlayerList(players: Player[], cap = 25): string {
  if (players.length === 0) return "No players found.";
  const shown = players.slice(0, cap);
  const lines = shown.map((p, i) => formatPlayer(p, i + 1));
  if (players.length > cap) {
    lines.push(`... (${players.length - cap} more players in dataset)`);
  }
  return lines.join("\n");
}
